statefun-sdk-js/src/core.ts

```diff
@@ -0,0 +1,329 @@
+        this.#namespace = namespace;
+        this.#name = name
+        this.#id = id;
+        this.#typename = typename;
+    }
+
+    /**
+     * Create an address that consist out of a namespace, name (also known as a typename) and an id.
+     *
+     * @param namespace the namespace part of the address
+     * @param name the name part of the address
+     * @param id the function's unique id.
+     * @returns {Address} an address that represents a specific function instance.
+     */
+    static fromParts(namespace: string, name: string, id: string) {
+        return new Address(namespace, name, id, `${namespace}/${name}`);
+    }
+
+    /**
+     * Creates an address from a <namespace>/<name> (aka typename) and an id pair.
+     * @param typename
+     * @param id
+     * @returns {Address}
+     */
+    // noinspection DuplicatedCode
+    static fromTypeNameId(typename: string, id: string) {
+        if (isEmptyOrNull(id)) {
+            throw new Error("id must be a defined string");
+        }
+        const {namespace, name} = parseTypeName(typename);
+        return new Address(namespace, name, id, typename)
+    }
+
+
+    /**
+     * @returns {string} returns the type name string (typename) of this function.
+     */
+    get typename() {
+        return this.#typename;
+    }
+
+    /**
+     * @returns {string} the id part of the address.
+     */
+    get id() {
+        return this.#id;
+    }
+
+    /**
+     *
+     * @returns {string} returns the namespace part of this address
+     */
+    get namespace() {
+        return this.#namespace;
+    }
+
+    /**
+     *
+     * @returns {string} returns the typename's name part of this address.
+     */
+    get name() {
+        return this.#name;
+    }
+}
+
+
+/**
+ * A representation of a single state value specification.
+ * This is created from the following object:
+ * {
+ *     name: string,
+ *     type: Type,
+ *     expireAfterCall / expireAfterWrite : int
+ * }
+ */
+export interface ValueSpecOpts {
+    name: string;
+    type: Type<any>;
+    expireAfterCall?: number;
+    expireAfterWrite?: number
+}
+
+/**
+ * A representation of a single function.
+ * This can be created with the following object:
+ * {
+ *     typename: "foo.bar/baz",
+ *     fn(context, message) {
+ *         ...
+ *     },
+ *     specs: [..]
+ * }
+ */
+export interface FunctionOpts {
+    typename: string;
+    fn: (context: Context, message: Message) => void | Promise<void>;
+    specs?: ValueSpecOpts[]
+}
+
+/**
+ * an internal representation of function spec
+ */
+class ValueSpec implements ValueSpecOpts {
+    readonly #name: string;
+    readonly #type: Type<any>;
+    readonly #expireAfterCall: number;
+    readonly #expireAfterWrite: number;
+
+    constructor(name: string, type: Type<any>, expireAfterCall?: number, expireAfterWrite?: number) {
+        this.#name = name;
+        this.#type = type;
+        this.#expireAfterCall = expireAfterCall || -1;
```

Comment:
Nowaday we use `expireAfterCall ?? -1` instead of `||` since `||` has broken semantics for values other than null/undefined in this context.